src/tests/test_types.cpp

```diff
@@ -264,4 +264,109 @@ INSTANTIATE_TEST_SUITE_P(TestStringSuite, TestString,
                                            TestStringParam("abc", ""),
                                            TestStringParam("bb", "aa")));
 
+using DatePtr = typename std::unique_ptr<std::byte[], mo::DateTypeDeleter>;
+
+DatePtr DatePtrFromString(std::string const& s, std::unique_ptr<mo::DateType> const& date_type) {
+    return {date_type->MakeValue(boost::gregorian::from_simple_string(s)),
+            date_type->GetDeleater()};
+}
+
+TEST(TestDateSuite, Compare) {
+    std::unique_ptr<mo::DateType> date_type(
+            mo::CreateSpecificType<mo::DateType>(mo::TypeId::kDate, true));
+    auto test = [date_type = std::move(date_type)](
+                        std::string const& date1, std::string const& date2, mo::CompareResult res) {
+        auto ptr1 = DatePtrFromString(date1, date_type);
+        auto ptr2 = DatePtrFromString(date2, date_type);
+        EXPECT_EQ(date_type->Compare(ptr1.get(), ptr2.get()), res);
+    };
+    test("2004-07-28", "2004-07-28", mo::CompareResult::kEqual);
+    test("2005-01-20", "2004-07-28", mo::CompareResult::kGreater);
+    test("2023-09-01", "2024-10-25", mo::CompareResult::kLess);
+    test("2023-10-28", "2023-10-28", mo::CompareResult::kEqual);
+    test("2203-09-01", "2024-10-25", mo::CompareResult::kGreater);
+    test("1991-11-20", "2200-07-28", mo::CompareResult::kLess);
+}
+
+TEST(TestDateSuite, Dist) {
+    std::unique_ptr<mo::DateType> date_type(
+            mo::CreateSpecificType<mo::DateType>(mo::TypeId::kDate, true));
+    auto test = [date_type = std::move(date_type)](std::string const& date1,
+                                                   std::string const& date2, long res) {
+        auto left_val = DatePtrFromString(date1, date_type);
+        auto right_val = DatePtrFromString(date2, date_type);
+        EXPECT_EQ(long(date_type->Dist(left_val.get(), right_val.get())), res);
```

Comment:
Avoid `Type(object)` syntax for casting, it is really confusing because such expression could mean a function call, a constructor call and a variable declaration depending on the context. 
Moreover, it's a C-style cast and Google Style Guide [forbids](https://google.github.io/styleguide/cppguide.html#Casting) them.